### DISCUSSION TITLE
M-Bus Protocol

### DISCUSSION THREAD

**Discussion by iot-sle** [arendst/Tasmota#15122] [Q&A] [answered]
this is not about Modbus. Story: I use a ABB B-series smart meter. The RS-485 connector is offering Modbus and EB-Bus protocol. Unfortunately it is in use by the photovoltaik equipment. There is an additional IR-Interface, which I try to use now. On this interface the modbus protocol is not available instead it is offering M-Bus protocol and EQ-Bus protocol. I don't find anything about EQ-Bus so I tried to use SML sensor in raw mode to get the values of M-Bus. I already spend some time and in result I'm able to send some Datagram correctly and receive some values. Do anyone have some experience in using M-Bus protocol? Especially how to handle "correctly" the handshake between the "result" datagrams? For the received values, I succeeded in reading values sending 2 byte and 4 byte. The next problem is the M-Bus protocol allows beside ASCII response values the values 4 digit BCD ( 2Byte) 6 digit BCD (3 Byte) 8 digit BCD (4 byte). I guess I can read them as word, and long word and convert them into variables correctly. But I can't see how to do that with 12 digit BCD ( 6 Byte). I'm short on adding bcd datatype to sensor53 (sml) code. The question is: Is there another possibility to go aheah, without do the development? (@gemu2015 What's your opinion?)

**Comment by gemu2015** [comment]
if you can catch the 12 digit bcd e.g. by two decoder lines with a 32 and 16 bit value, it would be very easy to convert this to bcd by scripting.  but it wouldn't be a big deal to add a bcd decoder to the sml driver if you would provide the actual data format.

**Comment by iot-sle** [comment]
Thanks for your quick reply. M-Bus is using following data formats in its datagrams:
beside formats already implemented by you ( byte, word, long word) 
64 bit Integer (8 bytes)
BCD-4  (2 bytes)
BCD-6  (3 bytes)
BCD-8  (4 bytes)
BCD-12  (6 bytes)
and variable length ASCII with length byte just in front of the string.
At the ABB - B series all values are LSB coded. As said above decoding with a trailing s marker as in ssSSssSSs is working correctly.
examples:

| Code        |  Hex (uu uu uu ....)    |  decimal value | 
|-----------|---------------------|-----------------|
| BCD-4      | 12 34                     | = 3412 | 
| BCD-6      | 01 12 34 56           | = 56341201 |  
| BCD-8      | 01 12 34 56 78      | = 7856341201 |  
| BCD-12    | 01 12 34 56 78 99 | = 997856341201 | 

for all values the count of decimals is given. by example the value has 2 as count of decimal
so the BCD-8 value 012345678 will be interpreted as 78563412,01 decimal

could see the converter functions for BCD bytes are part of the IRutils.cpp ( uint8ToBcd(..) and bcdToUint8(..)) 
I would speed up my "project" if you could implement the converting functions. I'm still fighting with the timing of the datagrams.
Thanks a lot.

**Reply by gemu2015** [reply]
bcd(sel val)
converts a number to (sel=1) or from (sel=0) bcd
since scripter uses single precision float it can only hold 24 bits, so only 6 digits are converted. you may have to split your numbers.

[xdrv_10_scripter.ino.zip](https://github.com/arendst/Tasmota/files/8239937/xdrv_10_scripter.ino.zip)

**Comment by iot-sle** [comment]
Documentation of MBus can be found here: [https://m-bus.com/assets/downloads/MBDOC48.PDF] M-Bus Docuementation
Page 39 description of data formats.

**Reply by barbudor** [reply]
Just for the sake of mentioning it, MODBUS being based on RS485 is a multi-device bus.
Nothing prevent you to connect tasmota to the bus to listen to frames sent from the meter.
Of course if resquest/response is  needed it may be more complicated.

**Reply by iot-sle** [reply]
yes, I know. I started with MODBUS, but I had no success. M-Bus needs some request and answers with serveral datagrams of max 256 bytes containing all values in one of the response datagram ( telegram ) . The problem is more that the decoding of BCD is not available in sml sensor now.

**Comment by iot-sle** [comment]
Thank you. I did a quick test confirm it’s working, as you described below.

 

Von: gemu ***@***.*** 
Gesendet: Sonntag, 13. März 2022 16:25
An: arendst/Tasmota
Cc: iot-sle; Author
Betreff: Re: [arendst/Tasmota] M-Bus Protocol (Discussion #15122)

 

bcd(sel val)
converts a number to (sel=1) or from (sel=0) bcd
since scripter uses single precision float it can only hold 24 bits, so only 6 digits are converted. you may have to split your numbers.

xdrv_10_scripter.ino.zip <https://github.com/arendst/Tasmota/files/8239937/xdrv_10_scripter.ino.zip>

**Reply by gemu2015** [reply]
bcd decoder added to sml driver
format bcdN    (N = 2..12) 

no number limit since SML uses double precision floats

[xsns_53_sml.ino.zip](https://github.com/arendst/Tasmota/files/8242549/xsns_53_sml.ino.zip)

**Comment by iot-sle** [comment]
Thank you for your effort. I’m sorry to answer only now. Unfortunately for me the bcdN part is not working. I took a look at your c code and don’t see anything wrong in the decoding function. May be you see the mistake, by looking at the results. I take only 1 example and show only relevant line

Script:

>M 2

+1, ….
+2, ….
….
***@***.***,ID,,ID,0 ; =sml[11] ID (BCD-8)
….
***@***.***,ID1,,ID-L,0 ; sml[23]
***@***.***,ID2,,ID-H,0 ; [24]
# 

>S
ABB_ID=bcd(0 sml[24])*10000+bcd(0 sml[23]) 

>W
ABB ID calculated:<td style="width:20px;white-space:nowrap">%ABB_ID%</td>

Result in Tasmota main windows:

ABB ID                     99

ABB ID calculated: 1250297

So the result of bcd8 is always 99 but the calculated result of the split and converted lines is fine!

**Reply by gemu2015** [reply]
please post your complete script instead of these fragments. i can not determine what might be wrong

**Comment by iot-sle** [comment]
Ok. I thought it would be easier for you only to see relevant lines. In my answer by email there where no *** in format. Here is the complete script, the meter 1 (ESY-SWS) is currently not connected! 
```
>D 80
send_Domoticz=0
SWS_Imp=0
SWS_Exp=0
ABB_Imp=0 
ABB_Exp=0
ABB_ID=0
ABB_Ges=0
ABB_FRQ=0
Gas=0
StromEG=0
Temp=0
cnt3=0
GPow=0
pD1=1
pD2=6
korr=0.00
butKorr=0
test=""

>B
=>sensor53 r

>BS
->displaymode 0
dt [zl1c1T][l1c10t][x1y10h128v0][O] 
->TouchNum 1
->TouchThres3 40

>M 2
+1,16,s,0,9600,ESY-SWS
+2,3,rE1,0,9600,ABB,1,10,1040105016,107b108b16,105b106b16
1,=h==============================
1,=h
1,77070100000009ff@#,ID,,ID,0 ;1
1,=h
1,77070100010800ff@1000,E Bezug,kWh,Imp,2 ;2
1,77070100020800ff@1000,E Einsp.,kWh,Exp,2 ;3
1,=h------aktuelle Leistungen-----
1,77070100100700ff@1,P Gesamt,W,P_tot,2 ;4
1,77070100240700ff@1,P L1,W,P_L1,2 ;5
1,77070100380700ff@1,P L2,W,P_L2,2 ;6
1,770701004c0700ff@1,P L3,W,P_L3,2 ;7
1,=h-----aktuelle Spannungen------
1,77070100200700ff@1,*,V,U_L1,1 ;8
1,77070100340700ff@1,*,V,U_L2,1 ;9
1,77070100480700ff@1,*,V,U_L3,1 ;10

2,081072bcd8@1,ID,,ID,0 ; 11 ID (BCD)

2,8E80C0408400bcd8@1,E netto Gesamt,kWh,Gesamt,0 ;12
2,0E8400bcd8@1,E Bezug,kWh,Imp,0 ;13 Bezug
2,8E408400bcd8@1,E Einsp.,kWh,Exp,0 ;14

2,04A900ssSSssSSs@100,P Gesamt,W,P_tot,2 ;15
2,04A9FF8100ssSSssSSs@100,P L1,W,P_L1,2 ; 16
2,04A9FF8200ssSSssSSs@100,P L2,W,P_L2,2 ;17
2,04A9FF8300ssSSssSSs@100,P L3,W,P_L3,2 ;18

2,04FDC8FF8100uuUUuuUUs@10,*,V,U_L1,1 ;19
2,04FDC8FF8200uuUUuuUUs@10,*,V,U_L2,1 ;20
2,04FDC8FF8300uuUUuuUUs@10,*,V,U_L3,1 ;21

2,0AFFE900bcd4@1,FreQ,Hz,F ;22

2,081072uuUU@1,*,,ID-L,0 ; 23 showing correct result
2,081072xxxxuuUU@1,*,,ID-H,0 ; 24 showing correct result
2,=h==============================
# 
>T
; get telemetry sensor values
Temp=DS18S20#Temperature

>S
SWS_Imp=sml[2]
SWS_Exp=sml[3]
ABB_Ges=sml[12];
ABB_Exp=sml[14];
;ABB_ID=sml[11];
ABB_ID=bcd(0 sml[24])*10000+bcd(0 sml[23])
ABB_FRQ=sml[22];

Gas=pc[1]
StromEG=pc[2]

if chg[StromEG]>0 
{
->power2 1
}

if upsecs%tper==0 
and 
send_Domoticz!=0
{

print Imp: %SWS_Imp% kWh Exp: %SWS_Exp% kWh StromEG: %3(StromEG/1000)% kWh Gas: %2(Gas/100)% m³ Temp: %Temp% °C Uptime: %upsecs% s Teleperiod: %tper% s
}

if upsecs%1==0
and 
pwr[3]==1
{ 
dt [Dp0] ;update display aus
dp2.1
dt [l1c1T][l1c10t][l1c16]%1Temp%~f8C
dp6.3
dt [l3c1]SWS Gas:%(Gas/100)% m
dt [l4c1]SWS Exp:%SWS_Exp% kWh 
dt [l5c1]SWS IMP:%SWS_Imp% kWh 
dt [l6c1]StromEG:%(StromEG/1000)% kWh 
dt [l7c1]ABB EXP:%ABB_Exp% kWh 
dt [l8c1]ABB IMP:%ABB_Imp% kWh 
dt [x123y12]3 
dt [d] ;update display
;print Display update um %tstamp%
;print %pD1% %pD2%
dp0
}
>P
GPow=pwr[1]
if chg[GPow]>0
and 
GPow==0
{
->counter1 +1
}
->Power3 %pwr[3]% 
>W
ABB ID calculated:<td style="width:20px;white-space:nowrap">%ABB_ID%</td> 
;ABB Bezug:<td style="width:20px;white-space:nowrap">%2(ABB_Imp/100)% kWh</td> 

```
Result:

![image](https://user-images.githubusercontent.com/74416643/158957225-359c277a-4158-441b-86d4-bcc20c953a3c.png)

**Reply by gemu2015** [reply]
ok, now it is clear what was wrong, try this bcd8 should work now
[xsns_53_sml.ino.zip](https://github.com/arendst/Tasmota/files/8303116/xsns_53_sml.ino.zip)

**Comment by iot-sle** [comment]
thank you so much for your effort. It's closer to the final now. But the values are still wrong.

![grafik](https://user-images.githubusercontent.com/74416643/159004934-b4f31840-175b-4694-80a6-f398b7031eac.png)

Correct value ABB ID is 01250297 same problem on ABB E Bezug it should be 00116865.

**Reply by gemu2015** [reply]
ok, next try
[xsns_53_sml.ino.zip](https://github.com/arendst/Tasmota/files/8304877/xsns_53_sml.ino.zip)

**Comment by iot-sle** [comment, marked as answer]
It's working fine now. Thank you so much.

**Reply by gemu2015** [reply]
btw since you are the first one to get an MBUS device to work with the SML driver, it would be nice if you would add an example to the docs.

**Comment by iot-sle** [comment]
I already planned to do that. But it’s not completely working now. I can only show the values returned by the device without a special selection.

Whatever I tried to get different values by a request according to the manual, I failed until now. I can reach by example DIF 03 Vif 84 (complete energy imported) because it’s part  of the RSP_UD telegram received. When I try to send a SND_UD telegram to receive by example the complete energy exported: DIF 8E 40 VIF 84 to my device 10  (68070768531051088E40840E16 or 68070768731051088E40842E16 )  It’s not returned in any telegram.  But the request telegram is correct because I received an E5 answer from the client. So I keep my hand on that part.

**Comment by iot-sle** [comment]
I added m-bus for abb to the documentation one week ago. 

https://github.com/tasmota/docs/pull/953

It seems that the pull request is not handled until now. 

At the same time the end of the file smart-meter-interfaced.md seems to be corrupted somehow:

**Comment by skupu** [comment]
Is there any chance to get support for wired bmeters watermeter rfm-mb-1?
